enigma/src/main/java/org/quiltmc/enigma/api/Enigma.java

```diff
@@ -128,6 +142,94 @@ public List<NameProposalService> getNameProposalServices() {
 		return proposalServices;
 	}
 
+	/**
+	 * {@return all registered read/write services}
+	 */
+	public List<ReadWriteService> getReadWriteServices() {
+		return this.services.get(ReadWriteService.TYPE);
+	}
+
+	/**
+	 * Gets all supported {@link FileType file types} for reading or writing.
+	 * @return the list of supported file types
+	 */
+	public List<FileType> getSupportedFileTypes() {
+		return this.getReadWriteServices().stream().map(ReadWriteService::getFileType).toList();
+	}
+
+	/**
+	 * Gets the {@link ReadWriteService read/write service} for the provided {@link FileType file type}.
+	 * @param fileType the file type to get the service for
+	 * @return the read/write service for the file type
+	 */
+	public Optional<ReadWriteService> getReadWriteService(FileType fileType) {
+		return this.getReadWriteServices().stream().filter(service -> service.getFileType().equals(fileType)).findFirst();
+	}
+
+	/**
+	 * Parses the {@link FileType file type} of the provided path and returns the corresponding {@link ReadWriteService read/write service}
+	 * @param path the path to analyse
+	 * @return the read/write service for the file type of the path
+	 */
+	public Optional<ReadWriteService> getReadWriteService(Path path) {
+		return this.parseFileType(path).flatMap(this::getReadWriteService);
+	}
+
+	/**
+	 * Determines the mapping format of the provided path. Checks all formats according to their {@link FileType} file extensions.
+	 * If the path is a directory, it will check the first file in the directory. For directories, defaults to the enigma mappings format.
+	 * @param path the path to analyse
+	 * @return the mapping format of the path
+	 */
+	public Optional<FileType> parseFileType(Path path) {
+		List<FileType> supportedTypes = this.getSupportedFileTypes();
+
+		if (Files.isDirectory(path)) {
+			try {
+				final AtomicReference<Optional<File>> firstFile = new AtomicReference<>();
+				Files.walkFileTree(path, new SimpleFileVisitor<>() {
+					@Override
+					public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+						super.visitFile(file, attrs);
+						if (attrs.isRegularFile()) {
+							firstFile.set(Optional.of(file.toFile()));
+							return FileVisitResult.TERMINATE;
+						}
+
+						return FileVisitResult.CONTINUE;
+					}
+				});
+
+				if (firstFile.get().isPresent()) {
```

Comment:
I think `#isPresent` could throw a NPE here, since `firstFile` doesn't have an initial value, so it's null by default